### PR TITLE
Adaptive bounds calculation

### PR DESCRIPTION
**Pull request by stepankuzmin**
This PR implements the adaptive bounds calculation proposed here https://github.com/mapbox/mapbox-gl-js/pull/12199

> Alternatively, we could do adaptive bounds calculation similar to the one happening in `tile_transform.js` for alternative projections — start with 4 corners, and delve recursively into midpoints while they stray too much from Euclidean midpoints. This might still produce a noticeable error, but will at least be a significant improvement to the current results. We should also be careful of the case where midpoint is in the middle but the line gets curvy on either side of the midpoint.

To calculate the globe bounds, it recursively iterates other midpoints of the canvas and raycasts the screen coordinates to the globe, expanding the bounds. It also expands the bounds if poles are visible.

## Before

Screen Shot 2022-10-13 at 15 57 55-before

## After

Screen Shot 2022-10-13 at 15 58 15-after

## Before

Screen Shot 2022-10-17 at 14 12 10

## After

Screen Shot 2022-10-17 at 14 12 25

## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] include before/after visuals or gifs if this PR includes visual changes
 - [x] write tests for all new functionality
 - [x] document any changes to public APIs
 - [x] manually test the debug page
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 - [x] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog>Improve the globe bounds calculation</changelog>`

### CODE REVIEW

File: src/geo/transform.js [hunk 1]
```diff
@@ -1344,6 +1344,79 @@ class Transform {
...
+
+        // we pick an error threshold for calculating the bbox that balances between performance and precision
+        const maxErr = s / 16;
+
+        function processSegment(pa, pb, ax, ay, bx, by) {
+            const mx = (ax + bx) / 2;
+            const my = (ay + by) / 2;
+
+            const pm = projection.project(lngFromMercatorX(mx), latFromMercatorY(my));
+            const err = Math.max(0, minX - pm.x, minY - pm.y, pm.x - maxX, pm.y - maxY);
```
Comment (stepankuzmin) [inline src/geo/transform.js:1391]: The `err` is always `0` here since the differences with the midpoint are always negative.

File: test/unit/ui/map.test.js [hunk 2]
```diff
@@ -1426,6 +1426,27 @@ test('Map', (t) => {
...
+            t.stub(console, 'warn');
+            const globeBounds = map.getBounds();
+
+            t.same(
+                toFixed(globeBounds.toArray()),
+                toFixed([[ -73.8873304141, 73.8873304141, ], [ 73.8873304141, -73.8873304141]])
+            );
+
+            t.end();
+        });
```
Comment (SnailBones) [inline test/unit/ui/map.test.js:1448]: Nit: Can we test a few more edge cases? Such as looking at the poles?

**Review by SnailBones — APPROVED** [review]
LGTM, nice work @stepankuzmin! Added one nit about testing coverage.

File: test/unit/ui/map.test.js (hunk 2 above)
Comment (stepankuzmin) [reply test/unit/ui/map.test.js:1448]: Sure!

File: src/geo/transform.js [hunk 3]
```diff
@@ -1344,7 +1344,60 @@ class Transform {
             new Point(Number.MAX_VALUE, Number.MAX_VALUE);
     }
 
+    _getGlobeBounds(): LngLatBounds {
+        const topLeft = new Point(this._edgeInsets.left, this._edgeInsets.top);
+        const topRight = new Point(this.width - this._edgeInsets.right, this._edgeInsets.top);
+        const bottomRight = new Point(this.width - this._edgeInsets.right, this.height - this._edgeInsets.bottom);
+        const bottomLeft = new Point(this._edgeInsets.left, this.height - this._edgeInsets.bottom);
```
Comment (mourner) [inline src/geo/transform.js:1351]: Nit: you can simplify this a bit by defining `top`/`bottom`/`left`/`right` variables instead and using them later:

```suggestion
        const {top, left} = this._edgeInsets;
        const bottom = this.height - this._edgeInsets.bottom;
        const right = this.width - this._edgeInsets.right;
```

File: src/geo/transform.js (hunk 1 above)
Comment (mourner) [reply src/geo/transform.js:1373]: The error metric here is a little unclear — can you explain what's measured here and when the recursion is expected to stop in practice, and add a comment perhaps?

**Review by mourner — COMMENTED** [review]
Nice work!

File: src/geo/transform.js (hunk 1 above)
Comment (stepankuzmin) [reply src/geo/transform.js:1372]: I'm not sure that it's the best approach here, but this is the same metric we use in `tileTransform`, and the recursion works the same way.

The `err` metric is the maximum distance between the midpoint and each of the currently calculated bounds. 

https://github.com/mapbox/mapbox-gl-js/blob/b7dc6dc935cc2c2abd3c70444f36118bcba7e522/src/geo/projection/tile_transform.js#L59

We expand the bounds on each iteration by the position of the midpoint, and the recursion stops when the error metric fall below the `maxErr` calculated based on the scale.

```js
const s = Math.pow(2, -zoom);
const maxErr = s / 16;
```

https://github.com/mapbox/mapbox-gl-js/blob/b7dc6dc935cc2c2abd3c70444f36118bcba7e522/src/geo/projection/tile_transform.js#L61-L69

**Review by mourner — APPROVED** [review]
